### ISSUE TITLE
Support new QUnit modules "acceptance testing"

### ISSUE THREAD

**Issue by bendemboski** [san650/ember-cli-page-object#355]
Using the new QUnit modules API, "acceptance tests" (which aren't exactly a thing anymore) don't register the global test helpers `visit`, `click`, etc. Instead such tests import the helpers explicitly from `ember-test-helpers`. Here's the blueprint:

```javascript
import { module, test } from 'qunit';
import { visit, currentURL } from '@ember/test-helpers';
import { setupApplicationTest } from 'ember-qunit';

module('blah blah blah', function(hooks) {
  setupApplicationTest(hooks);

  test('visiting /some-route', async function(assert) {
    await visit('/some-route');

    assert.equal(currentURL(), '/some-route');
  });
});
```

Currently we can't use a page objects because in acceptance test mode, the helpers are unavailable.

**Comment by bendemboski** [comment]
I looked into just modifying the existing acceptance context to use them, but `ember-test-helpers`'s helpers all rely on [getElement()](https://github.com/emberjs/ember-test-helpers/blob/master/addon-test-support/%40ember/test-helpers/dom/-get-element.js#L3) which doesn't accept a second context parameter (meaning none of the helpers do). Instead it uses the test context's element, which AFAIK is set (in acceptance tests) when `visit()` is called [here[(https://github.com/emberjs/ember-test-helpers/blob/master/addon-test-support/%40ember/test-helpers/setup-application-context.js#L16).

So I'm not quite sure what this means for `ember-cli-page-object` -- maybe in acceptance mode we just have to ignore the `testContainer` option?

I'm willing to spend some time contributing here unless you'd rather do it yourself, but need a little guidance 😄

**Comment by san650** [comment]
We need to start a conversation with the people behind `ember-test-helpers` to extend their helpers to receive an optional `context` param to override the use of the application roots element as the context.

https://github.com/emberjs/ember-test-helpers/blob/4fbc32f53c221cc4ac709ae9b99095197d6e3282/addon-test-support/%40ember/test-helpers/dom/-get-element.js#L12

The big problem we have to solve is how we're going to support the `testContainer` option from the page objects -- the option is used a lot for testing applications that use ember-wormhole and similar addons.

Maybe @rwjblue can give us his opinion on this problem, could we do a proposal for `ember-test-helpers` addon to support overriding the `context` element for `click`, `fillIn`, etc.?

@bendemboski We could start working on the migration to `ember-test-helpers` for acceptance tests in a separate PR, it would be great if you work on this, I can guide you on achieving this 😃

**Comment by bendemboski** [comment]
@san650 I had a chat @rwjblue this morning and think I understand how to achieve this. In a nutshell, we should just use DOM APIs for finding elements, and then pass those elements (instead of selectors) to the test helpers. I should have a PR to illustrate the direction later today.

**Comment by bendemboski** [comment]
I've run into a bit of a collision between existing `ember-cli-page-object` chaining syntax and the new test helpers. Kinda the main point of the new test helpers is to supply a consistent API across all kinds of tests, so acceptance and integration tests don't interact with the DOM in such syntactically different ways (in fact the distinction between acceptance and integration tests doesn't really exist anymore, they just are test modules set up with a different set of traits).

`ember-cli-page-object` sort of does that, but not exactly when it comes to asynchonicity. For example, consider the following code:

```javascript
page
  .clickOn('.open')
  .clickOn('.choose')
```

Assume that `.choose` isn't in the DOM to start with, and clicking on `.open` triggers an asynchronous operation that, when it completes, results in `.choose` being in the DOM. In an acceptance test, this would work just fine because `runAsync()` `wait()`s before running the operation. However, in an integration test this would fail because it doesn't do any automatic waiting.

This is not a criticism of `ember-cli-page-object` -- this is entirely in line with how the two different tests are expected to behave -- acceptance tests always wait for things to settle between helper executions, and integration tests require the user to explicitly call `wait()` to wait for things to settle.

The new test API takes the best of both -- you aren't forced to wait for things to settle between helper execution (so it's possible to test loading states and the like), but when you want to wait for things to settle, it's made very easy by the fact that all the helpers return a promise that ends with a call to `settled()` (the new `wait()`). So you can:

```javascript
click('.open'); // don't wait for things to settle
click('.choose');
```

or

```javascript
await click('.open'); // don't wait for things to settle
click('.choose');
```

or if you want to be more explicit, this is equivalent to the previous example:

```javascript
click('.open'); // don't wait for things to settle
await settled(); // now wait for things to settle
click('.choose');
```

We definitely want to maintain this flexibility when using `ember-cli-page-object` (imagine the above syntax except `await`ing page object calls instead), which means that the execution context for the new test API must not call `settled()` in `runAsync()` or you wouldn't be able to run two async operations without settling in between. Unfortunately, this means that chaining constructs like this will not work:

```javascript
page.visit()
  .clickOn('.open');
```

because the route transition will still be happening when the click operation executes. The same will be true of any other acceptance test chaining that relies on settling in between operations.

So, the options that I see are one or more of:

1. Document this and when users switch their acceptance tests over to the new modules API they will have to update any such chained invocations.
1. Add a `settled()` method to the base DSL structure so users can easily sprinkle chained calls to it where needed
1. Add a compatibility-mode method to opt in to always calling `settled()` from `runAsync()` (would have to be enabled in `beforeEach()` and disabled in `afterEach()` similar to setting the integration test context now).
1. Disable chaining when using the new test API execution context since it's kinda confusing (although this would be a breaking change to any integration tests using `ember-cli-page-object` that have already been ported to the new test API and use chaining).

@san650, thoughts?

**Comment by magistrula** [comment]
Just to throw out another option:

5. In acceptance tests, allow an options hash to be passed into `ember-cli-page-object` helpers that specifies whether the helper should call `settled` or not, with the default being the current behavior, that `settled` **is** called.
```javascript
// Visit page and wait for things to settle
// Click "open" but don't wait for things to settle
page.visit()
  .clickOn('.open', { await: false })
```

```javascript
// Visit page and wait for things to settle
// Click "open" and do wait for things to settle
page.visit()
  .clickOn('.open')
```

**Comment by bendemboski** [comment]
Yeah, that would work too. My personal preference would be to make waiting opt-in rather than opt-out because that's the way the new test API works. For example, I'd prefer to see

```javascript
await visit('/page');
click('.foo');
click('.bar');
```

be the equivalent of

```javascript
await page.visit();
page.clickFoo();
page.clickBar();
```

rather than

```javascript
await page.visit();
page.clickFoo();
page.clickBar({ await: false });
```

or the equivalent

```javascript
page.visit();
page.clickFoo();
page.clickBar({ await: false });
```

and I'm not sure how to implement the chaining syntax/behavior you describe without also forcing this departure from the normal test syntax when not chaining.

From a higher level, I see `ember-cli-page-object` as providing two pieces of value -- providing a DSL, and providing a consistent API for various types of tests. Now that the core Ember test framework is providing a consistent API for all test types, I don't see that as `ember-cli-page-object`'s job anymore, so it should just be focused on providing a DSL and being completely agnostic as to async/wait behavior, just passing through the behavior of the underlying helpers. Since the underlying helpers require an explicit `await` call to opt in to waiting, I think `ember-cli-page-object` should as well.

But I'm a newcomer here, and I don't have any ego invested, so if this doesn't match the philosophy/direction of the addon, I'll happy capitulate 😄

**Comment by ro0gr** [comment]
First of all I'm very excited to have new testing API support in `ember-cli-page-object`!

@bendemboski I think there is one more option to consider. It could potentially allow us to have a better backward compatibility story. 

What if we implement a new execution context using new test helpers and with the `runAsync` implemented like:
```js
  runAsync(cb) {
      cb(this);
      return settled().then(() => this.pageObjectNode);
  }
```
This way async helpers(`clickOnText`, `fillable`, etc) won't be forced to wait for async work to be done before be triggered. I believe it could allow us to preserve chainability. At the same time it should be also possible to use explicit `async`/`await`s. Thoughts?

**Comment by bendemboski** [comment]
@ro0gr unfortunately that wouldn't allow chaining because the return value of the chainable calls would be a promise rather than the original tree node (even though it would resolve to the original tree node).

I've put some more careful thought into this, and here's my analysis.

## The problem

I don't know of a way to maintain the current chaining async behavior _and_ reflect the new test API's async behavior without some trickery that I'm not sure how to accomplish (see below). Here's the problem.

Consider a chaining scenario:

```javascript
page
  .clickOn('.foo')
  .clickOn('.bar')
```

To preserve backwards compatibility, we'd need to wait for things to settle between the two `clickOn` calls. But the above is the equivalent of:

```javascript
page.clickOn('.foo');
page.clickOn('.bar');
```

which is the syntax-equivalent of the new testing API's

```javascript
// import { click } from '@ember/test-helpers';

click('.foo');
click('.bar');
```

which does not wait between the click calls unless the first click is `await`ed. So I don't see how it's possible to maintain async-consistency with current chaining _and_ the new test API's syntax because they are fundamentally incompatible. Unless...

## The trickery (a potential solution)

I don't think we can have it both ways _unless_ we break the equivalency between the first two examples, meaning that

```javascript
page.foo() !== page
```

In other words, if `page.foo()` did not wait, but instead of returning `page`, returned an object with all the same methods as `page`, but when invoked they wait before calling their helper (if async), we could get both behaviors. I don't know enough about `Ceibo` to know if it would be feasible, but, in pseudo-ish-code, it would be something like:

```javascript
runAsync(fn) {
  let promise = fn(this);
  let obj = {};
  this.pageObjectNode.forEachMethodName((key) => {
    let callThrough = (...args) => this.pageObjectNode[key].apply(this.pageObjectNode, args);

    if (method.isAsync) {
      obj[key] = (...args) => promise.then(() => callThrough(...args));
    } else {
      obj[key] = (...args) => callThrough(...args);
    }
  });
  return obj;
}
```

This would allow the behavior of sequential unchained method calls to mimic the new test API's async behavior (no waiting), and chained method calls to mimic the existing behavior (waiting). Assuming this is feasible...

## My opinion

I think that with the syntactic simplification of the new testing API (no nested functions resulting from promise chaining) chaining doesn't really gain us anything. Given the complexity of implementing it _and_ the fact that it would technically break compatibility one way or the other (either it doesn't wait and is incompatible with existing acceptance tests, or it does wait and is incompatible with existing integration tests), I think we should just abandon it. This would break compatibility with existing tests that do chaining, but the code updates would be entirely mechanical (maybe we could even write a codemod!).

**Comment by bendemboski** [comment]
BTW, if anybody's interested, [here](https://github.com/san650/ember-cli-page-object/compare/master...bendemboski:rfc268)'s my branch with all the changes to support the new testing API, but without changes to support or pull support for chaining (so all the new-test-api chaining unit tests are failing), and [here](https://github.com/san650/ember-cli-page-object/compare/master...bendemboski:rfc268-no-chaining)'s a branch with one additional commit that updates the unit tests to not expect the new test API to support chaining (and all the unit tests pass).

I'll probably open some sort of WIP PR tomorrow...

**Comment by juanazam** [comment]
When we first implemented chaining, it was a way of simplifying testing code to make it more readable and compact. Until this point, it hasn't been a lot of effort to keep it working version through version.

When we separated execution into contexts (acceptance and integration) we knew that at some point the work we had done would be included in the new ember testing solution.

I think the cost of maintaining chaining is now too high and we are moving away from the ember standard. We should evaluate removing it and making sure migrating away from it is as painless as possible.

That's my two cents

**Comment by bendemboski** [comment]
PR: https://github.com/san650/ember-cli-page-object/pull/356

**Comment by ro0gr** [comment]
TBH I'm not very happy to drop a chaining support. In my opinion it's quite convenient feature.
I believe in tests most of the time we need an async behavior so if we loose chaining we might end up with endless `await`s. Let's compare a synthetic example with two syntaxes:

- chaining
```js 
 await registrationPage
    .fillName('Name')
    .fillSurname('Surname')
    .fillPhone('+xxxxxxxxxx')
    .fillPassword('xxxxxxxxx')
    .submit();
```
- async/await
```js
  await registrationPage.fillName('Name');
  await registrationPage.fillSurname('Surname');
  await registrationPage.fillPhone('+xxxxxxxxxx');
  await registrationPage.fillPassword('xxxxxxxxx');
  await registrationPage.submit();
```

I find that the chaining syntax is much less verbose. In the second example we're forced to repeat `await registrationPage.` part all the time.

My second point is that if we break chaining API we will force users to update their test suites or stay on outdated `e-c-p-o` version. I agree that codemod can make the magic but I still believe if we can add this feature w/o breakage of existing test suites it would allow to migrate existing test suites to the new test-helpers API more gradually.

---

@bendemboski I think potential solution you proposed(https://github.com/san650/ember-cli-page-object/issues/355#issuecomment-354064968) worth to be explored. 
Do you have any luck with it?

**Comment by bendemboski** [comment]
## Syntax verbosity/[brevity](https://www.youtube.com/watch?v=Be7Og9Gc_KY)

@ro0gr agreed that dropping chaining forces a more verbose syntax. It occurs to me that with a very small code change we could support a more concise form of your example.

```javascript
await registrationPage.as((r) => {
  await r.fillName('Name');
  await r.fillSurname('Surname');
  await r.fillPhone('+xxxxxxxxxx');
  await r.fillPassword('xxxxxxxxx');
  await r.submit();
});
```

Although it's still not as concise as chaining. However, whether or not we support chaining, when _not_ chaining, I strongly urge us to support the async/await syntax in such a way as to reflect the RFC268 test syntax. I believe this is the right decision for a number of reasons, not the least of which is that in the future when everybody writes RFC268 tests, newcomers to `ember-cli-page-object` will have a much easier learning curve. Does anybody disagree with that?

If everybody agrees there, then the question is how much effort is it to support that _and_ chaining, and is that effort worth the IMO (now that the base test API is nice and consistent) fairly modest syntactic/ergonomic gains.

## Migration & compatibility

No matter what we do with the RFC268 execution context, no existing test suites will break. Integration tests, and their RFC268 successor, rendering tests, work using the integration execution context, which is enabled by calling `page.setContext(this)`, and this is the only way to get `ember-cli-page-object` working with rendering tests. The new execution context switching code looks like this in pseudo-code:

```javascript
if (wasSetContextCalled()) {
  return new IntegrationContext();
} else if (isRfc268API() {
  return new Rfc268Context();
} else {
  return new AcceptanceContext();
}
```

So any existing tests using `ember-cli-page-object` will continue to execute with their current (legacy) execution context. In order to start using the RFC268 execution context, the developer would have to:

1. For integration tests, migrate to RFC268 `setupRenderingTest()` tests and delete the `page.setContext(this)`
1. For acceptance tests, migrate to RFC268 `setupApplicationTest()` tests
1. For RFC268 rendering tests that are already using `ember-cli-page-object` via the integration context, delete the `page.setContext(this)`.

The above can be done module by module, so my entire PR is a non-breaking change as it requires positive action to start using the new execution context. So we can implement whatever behavior we want without it being a breaking change.

## Supporting chaining

My previous idea for allowing chained calls to differ in their async behavior from unchained calls didn't work out (I can't figure out the difference between calls that do and don't need to be wrapped in waiting behavior when cloning a node). So my latest idea is at tree construction time, to try to also construct a mirror of the tree where the nodes all have a property set that distinguish them from the primary tree nodes. Then any async calls can return their mirror node, and chained calls to methods on that node (or any of its descendant nodes) would be identifiable to the execution context so it would know to block before invoking the method.

I'll give this a shot, but if it doesn't work, I'm kinda out of ideas and either need somebody that knows `Ceibo` better than me to take a look, or just give up on the whole direction, which I think means abandoning chaining.

**Comment by bendemboski** [comment]
🎉 my last idea for how to support the pre-existing chaining async behavior worked. It's in its own commit in the PR I just opened (#357), and isn't too terribly hacky 😄 Other than documentation and maybe improving test coverage a tiny bit, I think that PR is ready to go, so anybody that wants to look over the code or try it out with your test suite, have at it!

**Comment by Sinled** [comment]
Any news on this issue? New tests api is already here, but i cannot migrate my tests yet, because of incompatibility.

**Comment by ro0gr** [comment]
@Sinled I think https://github.com/san650/ember-cli-page-object/pull/357 is in a pretty good shape. It's going to be included in the 2.0 release. I think we still need some docs.
